google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/Query.java

```diff
@@ -147,6 +154,76 @@ public Query limit(long limit) {
     return this;
   }
 
+  /**
+   * Split this query into multiple queries that can be evenly distributed across Bigtable nodes and
+   * be run im parallel. This method takes the results from {@link
+   * com.google.cloud.bigtable.data.v2.BigtableDataClient#sampleRowKeysAsync(String)} to divide this
+   * query into a set of disjoint queries that logically combine into form this query.
+   *
+   * <p>Expected Usage:
+   *
+   * <pre>{@code
+   * List<KeyOffset> keyOffsets = dataClient.sampleRowKeysAsync("my-table").get();
+   * List<Query> queryShards = myQuery.shard(keyOffsets);
+   * List<ApiFuture<List<Row>>> futures = new ArrayList();
+   * for (Query subQuery : queryShards) {
+   *   futures.add(dataClient.readRowsCallable().all().futureCall(subQuery))
+   * }
+   * List<List<Row>> results = ApiFutures.allAsList(futures).get();
+   * }</pre>
+   */
+  public List<Query> shard(List<KeyOffset> sampledRowKeys) {
+    Preconditions.checkState(builder.getRowsLimit() == 0, "Can't shard query with row limits");
+
+    ImmutableSortedSet.Builder<ByteString> splitPoints =
+        ImmutableSortedSet.orderedBy(ByteStringComparator.INSTANCE);
+
+    for (KeyOffset keyOffset : sampledRowKeys) {
+      if (!keyOffset.geyKey().isEmpty()) {
+        splitPoints.add(keyOffset.geyKey());
+      }
+    }
+
+    return shard(splitPoints.build());
```

Comment:
Unless we need to continue adding elements in there to be sorted, we perhaps to convert it  to a list.

Reply:
Please see the comment for `getRemaining`, but the short version is that I want the input be pre-sorted and SortedSet allows me to communicate that to the caller

Reply:
Well, by converting to a list, you get to keep the sorted order, no?